### ISSUE TITLE
xe_ndcg implementation is different from the paper

### ISSUE THREAD

**Issue by we1559** [microsoft/LightGBM#3673]
hi, @sbruch I'm working with the xe_ndcg now, but I don't understand the second order's meaning, 
https://github.com/microsoft/LightGBM/blob/master/src/objective/rank_objective.hpp#L339

so I refer to the original pr which you create
https://github.com/microsoft/LightGBM/pull/2620/files#diff-0d684c70b87b4c45b2e7054d75fd3e82324c3cff050fdaf79c4545c23642c496R107

the paper said the gradient should be as follow:
g(i) = -phi(i) + rho(i)

but your implementation is 
g(i) = -phi(i) + rho(i) + [rho[i]*L2s[i] + rho[i]*L3s[i])]

I'm confused aboud the last two item.

would you explain it for me please, thank you.

**Comment by guolinke** [comment]
cc @sbruch

**Comment by sbruch** [comment]
Please see Section 4.3 of the [paper](https://arxiv.org/pdf/1911.09798.pdf). The `lambdas` and `hessian` variables in this package correspond to the numerator and the denominator (respectively) of the last equation in that section.

**Comment by we1559** [comment]
Thank you for your replay.

The last equation in Section 4.3 is approximating the inverse of H. And we get H^(-1) = `lambdas` / `hessian`

But I think we only need to calculate the original gradient and hessian, the GBM could use them to get a target of this iteration and update.

So I'm still confused. Why we need to update gradient's calculation according to the inverse of H?

**Comment by sbruch** [comment]
I'm not sure what's the source of confusion, but I'm perhaps not explaining this well. The optimization procedure in LightGBM is [Newton's second order method](https://en.wikipedia.org/wiki/Newton%27s_method_in_optimization). That is what requires the Hessian, in addition to the gradient, to compute an update direction.

The last equation in Section 4.3 is not just the inverse of Hessian, but the Newton's update rule.

Unfortunately, the LightGBM API does not accommodate multivariate functions well: it doesn't allow one to specify an inverted Hessian matrix (H^{-1}) and gradient (\nabla). It takes instead some degenerate "Hessian" per output dimension. That sort of API works well for functions with a simple Hessian like LambdaMART, but the same procedure does not work for more complex ones like Softmax. So I instead hijacked `lambdas` and `hessian` in a way that would ultimately produce update rules according to H^-1 * \nabla. (Without regularization, that's essentially the update rule you'll get given `lambdas` and `hessian` variables.) That's why `lambdas` and `hessian` are set to what they are in the code: H^{-1} * \nabla becomes (in its simplest form) `lambdas` / `hessian`.

Again, of course, ideally one should just provide H^{-1} and \nabla and calculate the update rule for every output dimension. Alas that can't be done here. But if/when that's feasible, that would be the correct implementation.

**Comment by we1559** [comment]
> That sort of API works well for functions with a simple Hessian like LambdaMART, but the same procedure does not work for more complex ones like Softmax
Maybe that's what I'm confused about. Is there any assumption about objective function in LightGBM？

**Comment by we1559** [comment]
Maybe I get the result.

> That sort of API works well for functions with a simple Hessian like LambdaMART, but the same procedure does not work for more complex ones like Softmax

the simple Hessian you mentioned above is a diagonal matrix, so it's inverted matrix is H^-1 ii =  1 / H ii, H^-1 ij = 0(if i != j).
Finally used it to calculate the leaf value here https://github.com/microsoft/LightGBM/blob/master/src/treelearner/feature_histogram.hpp#L750

But in the complex Hessian situation, the inverted Hessian is not a diagonal matrix, so the code above should change, but you just hijacked `lambdas` and `hessian` to avoid changing the framework.

Am I right?

**Comment by sbruch** [comment]
> the simple Hessian you mentioned above is a diagonal matrix, so it's inverted matrix is H^-1 ii = 1 / H ii, H^-1 ij = 0(if i != j).

Yes, precisely.

> But in the complex Hessian situation, the inverted Hessian is not a diagonal matrix, so the code above should change, but you just hijacked `lambdas` and `hessian` to avoid changing the framework.
> 
> Am I right?

Correct.

**Comment by we1559** [comment]
Thanks for your explaination
